### ISSUE TITLE
Cython generates unsafe None-tolerant code

### ISSUE THREAD

**Issue by pitrou** [cython/cython#2170]
With Cython 0.28.1. Given the following method:
```python
    @staticmethod
    def from_buffers(DataType type, length, buffers, null_count=-1, offset=0):
        cdef:
            Buffer buf
            vector[shared_ptr[CBuffer]] c_buffers
            shared_ptr[CArrayData] ad

        for buf in buffers:
            c_buffers.push_back(buf.buffer)
        # [snip rest of method]
```

Cython generates the following C snippet for the `for` loop:
```c++
  struct __pyx_obj_7pyarrow_3lib_Buffer *__pyx_v_buf = 0;
  /* ... */
  for (;;) {
  /* ... */
    if (!(likely(((__pyx_t_4) == Py_None) || likely(__Pyx_TypeTest(__pyx_t_4, __pyx_ptype_7pyarrow_3lib_Buffer))))) __PYX_ERR(2, 403, __pyx_L1_error)
    __Pyx_XDECREF_SET(__pyx_v_buf, ((struct __pyx_obj_7pyarrow_3lib_Buffer *)__pyx_t_4));
    __pyx_t_4 = 0;

    /* "../../pyarrow/array.pxi":404
 * 
 *         for buf in buffers:
 *             c_buffers.push_back(buf.buffer)             # <<<<<<<<<<<<<<
 *         ad = CArrayData.Make(type.sp_type, length, c_buffers,
 *                              null_count, offset)
 */
    try {
      __pyx_v_c_buffers.push_back(__pyx_v_buf->buffer);
    } catch(...) {
      __Pyx_CppExn2PyErr();
      __PYX_ERR(2, 404, __pyx_L1_error)
    }
```

As you see `Py_None` is allowed and then blindly casted to a `struct __pyx_obj_7pyarrow_3lib_Buffer *`, from which the `buffer` member is then dereferenced.

I would expect Cython to either refuse None here (letting me implement the required logic myself) or check for None when dereferencing an attribute (and raise an exception perhaps?).

**Comment by scoder** [comment]
I agree that this can be surprising, but it's been like that ... maybe forever? I'm pretty sure that behaviour is documented somewhere.

The fact that None is allowed is ok, since all object typed variables necessarily support a None assignment. The cast is required afterwards to support the assignment of PyObject* to the cdef class typed variable.

The thing then is that accessing fields of cdef classes does not check for None, and that's for two reasons: performance, and to make them accessible inside of nogil sections. There is a somewhat hand-waving argument that accessing a cdef class field is an unsafe C operation and not a Python protocol, a bit of "practicality beats purity".

I could imagine issuing an invisible warning for this, but a) that means that people still wouldn't normally see it, and b) if they increase the warning level, they'd probably be swamped with false positives, unless we implement very good None check tracing, which we currently do not do at all in the control flow analysis (we leave that to the C compiler, which does a very good job eliminating our generated None checks).

I'm leaving this open as it might actually be worth revisiting the impact of this with recent compilers and CPUs, although that doesn't resolve the problem of breaking existing working (and correct) code that does not support raising an exception from a given function.

**Comment by pitrou** [comment]
I see the "nonecheck" compiler directive which looks like a good enough solution for us.
